src/java/htsjdk/samtools/CRAMIndexer.java

```diff
@@ -101,16 +105,72 @@ public CRAMIndexer(final OutputStream output, final SAMFileHeader fileHeader) {
     }
 
     /**
-     * Record any index information for a given CRAM slice.
+     * Index a container, any of mapped, unmapped and multiple references are allowed. The only requirement is sort
+     * order by coordinate.
+     * For multiref containers the method reads the container through unpacking all reads. This is slower than single
+     * reference but should be faster than normal reading.
+     *
+     * @param container container to be indexed
+     */
+    public void processContainer(final Container container) {
+        try {
+            if (container == null || container.isEOF()) {
+                return;
+            }
+
+            int sliceIndex = 0;
+            for (final Slice slice : container.slices) {
+                slice.containerOffset = container.offset;
+                slice.index = sliceIndex++;
+                if (slice.isMultiref()) {
+                    final ContainerParser parser = new ContainerParser(indexBuilder.bamHeader);
+                    final Map<Integer, AlignmentSpan> refSet = parser.getReferences(container);
+                    final Slice fakeSlice = new Slice();
+                    slice.containerOffset = container.offset;
+                    slice.index = sliceIndex++;
+                    for (final int refId : refSet.keySet()) {
+                        fakeSlice.sequenceId = refId;
+                        final AlignmentSpan span = refSet.get(refId);
+                        fakeSlice.containerOffset = slice.containerOffset;
+                        fakeSlice.offset = slice.offset;
+                        fakeSlice.index = slice.index;
+
+                        fakeSlice.alignmentStart = span.getStart();
+                        fakeSlice.alignmentSpan = span.getSpan();
+                        fakeSlice.nofRecords = span.getCount();
+                        processSingleReferenceSlice(fakeSlice);
+                    }
+                } else {
+                    processSingleReferenceSlice(slice);
+                }
+            }
+
+        } catch (final IOException e) {
+            throw new RuntimeIOException("Failed to read cram container", e);
+        } catch (final IllegalAccessException e) {
+            throw new SAMException("Failed to read cram container", e);
+        }

```

Comment:
It looks like these exceptions (i.e. the IllegalAccessException) originate deep down at the encoding level, and are propagated up through several layers as checked exceptions, but they aren't really handled anywhere. I think it would be a lot more useful to catch these where they originate (DataReader/Factory ??) and translate them into unchecked exceptions with a more meaningful message saying what went wrong. Then all the other layers wouldn't have to declare and catch them, since they can't do anything with them anyway.